### PR TITLE
Refactor window operations to avoid WindowConfigImpl cast

### PR DESCRIPTION
**Pull request by nwangtw**
- The WindowConfigImpl has a type variable and a select block. Refactor them into separated class to make the code more readable.
- Add an attachWindowConfig() function in WindowConfig interface to avoid casting in code.

### CODE REVIEW

File: heron/api/src/java/org/apache/heron/streamlet/WindowConfig.java [hunk 1]
```diff
@@ -26,7 +26,10 @@
 import org.apache.heron.api.tuple.Tuple;
 import org.apache.heron.api.windowing.EvictionPolicy;
 import org.apache.heron.api.windowing.TriggerPolicy;
-import org.apache.heron.streamlet.impl.WindowConfigImpl;
+import org.apache.heron.streamlet.impl.windowings.CountWindowConfig;
+import org.apache.heron.streamlet.impl.windowings.CustomWindowConfig;
+import org.apache.heron.streamlet.impl.windowings.TimeWindowConfig;
+//import org.apache.heron.streamlet.impl.WindowConfigImpl;
```
Comment (xiaoyao1991) [inline heron/api/src/java/org/apache/heron/streamlet/WindowConfig.java]: unused import

File: heron/api/src/java/org/apache/heron/streamlet/impl/WindowConfigImpl.java [hunk 2]
```diff
@@ -19,59 +19,12 @@
...
-        break;
-      case CUSTOM:
-        bolt.withCustomEvictor(evictionPolicy);
-        bolt.withCustomTrigger(triggerPolicy);
-        break;
-      default:
-        throw new RuntimeException("Unknown windowType " + String.valueOf(windowType));
-    }
-  }
+public abstract class WindowConfigImpl implements WindowConfig {
```
Comment (xiaoyao1991) [inline heron/api/src/java/org/apache/heron/streamlet/impl/WindowConfigImpl.java]: Since it's no longer an Impl class, could we move the abstract method to the interface itself and get rid of this class?

File: heron/api/src/java/org/apache/heron/streamlet/impl/streamlets/GeneralReduceByKeyAndWindowStreamlet.java [hunk 3]
```diff
@@ -66,7 +65,7 @@ public boolean doBuild(TopologyBuilder bldr, Set<String> stageNames) {
     setDefaultNameIfNone(StreamletNamePrefix.REDUCE, stageNames);
     GeneralReduceByKeyAndWindowOperator<K, V, VR> bolt =
         new GeneralReduceByKeyAndWindowOperator<K, V, VR>(keyExtractor, identity, reduceFn);
-    windowCfg.attachWindowConfig(bolt);
+    bolt.applyWindowConfig(windowCfg);
```
Comment (xiaoyao1991) [inline heron/api/src/java/org/apache/heron/streamlet/impl/streamlets/GeneralReduceByKeyAndWindowStreamlet.java]: `applyWindowConfig` and `attachWindowConfig` seem redundant to each other. We could call `windowCfg.attachWIndowConfig(bolt)` here directly right?

File: heron/api/src/java/org/apache/heron/streamlet/impl/WindowConfigImpl.java (hunk 2 above)
Comment (nwangtw) [reply heron/api/src/java/org/apache/heron/streamlet/impl/WindowConfigImpl.java]: I tried and it doesn't work. Let me try again.

File: heron/api/src/java/org/apache/heron/streamlet/WindowConfig.java (hunk 1 above)
Comment (nwangtw) [reply heron/api/src/java/org/apache/heron/streamlet/WindowConfig.java]: will do. thx.

File: heron/api/src/java/org/apache/heron/streamlet/impl/streamlets/GeneralReduceByKeyAndWindowStreamlet.java (hunk 3 above)
Comment (nwangtw) [reply heron/api/src/java/org/apache/heron/streamlet/impl/streamlets/GeneralReduceByKeyAndWindowStreamlet.java]: windowCfg doesn't have the function and requires a static cast. :(

File: heron/api/src/java/org/apache/heron/streamlet/impl/WindowConfigImpl.java (hunk 2 above)
Comment (nwangtw) [reply heron/api/src/java/org/apache/heron/streamlet/impl/WindowConfigImpl.java]: examples/src/java/org/apache/heron/examples/streamlet/ImpressionsAndClicksTopology.java:177: error: reference to reduceByKeyAndWindow is ambiguous
        .reduceByKeyAndWindow(
        ^
  both method <K#1,V>reduceByKeyAndWindow(SerializableFunction<R,K#1>,SerializableFunction<R,V>,WindowConfig,SerializableBinaryOperator<V>) in Streamlet a
nd method <K#2,T>reduceByKeyAndWindow(SerializableFunction<R,K#2>,WindowConfig,T,SerializableBiFunction<T,R,? extends T>) in Streamlet match
  where K#1,V,R,K#2,T are type-variables:
    K#1 extends Object declared in method <K#1,V>reduceByKeyAndWindow(SerializableFunction<R,K#1>,SerializableFunction<R,V>,WindowConfig,SerializableBinar
yOperator<V>)
    V extends Object declared in method <K#1,V>reduceByKeyAndWindow(SerializableFunction<R,K#1>,SerializableFunction<R,V>,WindowConfig,SerializableBinaryOperator<V>)
    R extends Object declared in interface Streamlet
    K#2 extends Object declared in method <K#2,T>reduceByKeyAndWindow(SerializableFunction<R,K#2>,WindowConfig,T,SerializableBiFunction<T,R,? extends T>)
    T extends Object declared in method <K#2,T>reduceByKeyAndWindow(SerializableFunction<R,K#2>,WindowConfig,T,SerializableBiFunction<T,R,? extends T>)

File: heron/api/src/java/org/apache/heron/streamlet/impl/streamlets/GeneralReduceByKeyAndWindowStreamlet.java (hunk 3 above)
Comment (nwangtw) [reply heron/api/src/java/org/apache/heron/streamlet/impl/streamlets/GeneralReduceByKeyAndWindowStreamlet.java]: Cleaned up a bit more. Moved the call into Operator classes so we dont need to worry about the details here.

File: heron/api/src/java/org/apache/heron/streamlet/impl/WindowConfigImpl.java (hunk 2 above)
Comment (xiaoyao1991) [reply heron/api/src/java/org/apache/heron/streamlet/impl/WindowConfigImpl.java]: Hmm, looks like the compiler can't tell the difference between an interface and lambda expressions at compile time. We can combine `WindowConfig` and `WindowConfigImpl` still as one abstract class to remove redundancy.

File: heron/api/src/java/org/apache/heron/streamlet/impl/streamlets/GeneralReduceByKeyAndWindowStreamlet.java (hunk 3 above)
Comment (xiaoyao1991) [reply heron/api/src/java/org/apache/heron/streamlet/impl/streamlets/GeneralReduceByKeyAndWindowStreamlet.java]: > windowCfg doesn't have the function and requires a static cast. :(

hmm, if we combine `WindowConfig` and `WindowConfigImpl` into one class as an abstract class, then `windowCfg.attachWindowConfig(bolt)` should work.

Comment (nwangtw) [reply heron/api/src/java/org/apache/heron/streamlet/impl/streamlets/GeneralReduceByKeyAndWindowStreamlet.java]: Yeah. Thanks for figuring out the cause of the compiling error when the attachWindowConfig() is in the interface.

Let me update it.

**Comment by xiaoyao1991** [comment]
👍

File: heron/api/src/java/org/apache/heron/streamlet/impl/WindowConfigImpl.java (hunk 2 above)
Comment (nlu90) [reply heron/api/src/java/org/apache/heron/streamlet/impl/WindowConfigImpl.java]: +1 for reducing the redundancy.
